crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -391,6 +403,23 @@ impl<'db> UseDefMap<'db> {
             .is_always_false()
     }
 
+    pub(crate) fn is_attribute_assignment_visible(
+        &self,
+        db: &dyn crate::Db,
+        symbol: ScopedSymbolId,
+        id: ScopedDefinitionId,
+    ) -> Truthiness {
+        let visibility = self
+            .attribute_assignments(symbol)
+            .find_map(|bind| (id == bind.definition_id).then_some(bind.visibility_constraint));
+        let Some(visibility) = visibility else {
+            // Overwritten (e.g. `self.x = 1; self.x = 2`)
```

Comment:
This is an improvement that we should probably add a test for?

```py
class C:
    def __init__(self):
        self.x = 1
        self.x = "a"

reveal_type(C().x)
```

Previously, we would infer `Unknown | Literal[1, 2]` in that case, and now we infer the more correct `Unknown | Literal[2]`.